File: make_past_data.pl
#!/usr/bin/env perl
use utf8;
use open qw(:std :utf8);
use autodie;
use Encode;
use strict;
use warnings;

use Config::Simple;

my $cfg = new Config::Simple('config.ini');
my $currency = $cfg->param('settings.currency_pair');

if (scalar(@ARGV) != 2) {
    print STDERR "$0 <start week> <end week>\n";
    exit(-1);
}

my ($start_week, $end_week) = @ARGV;
my @test_files;
while (<$currency/weekly_past_data/week_*.csv>) {
    m{week_(\d{3})};
    next unless $1 >= $start_week and $1 <= $end_week;
    push @test_files, $_;
}

my $past_width = 120000;
my $future_width = 30000;
my $count = 1;

open my $fh_out_result, ">", "$currency/past_data.txt";
my $week_num = 1;
for my $test_file(@test_files) {
    print STDERR "$test_file\n";
    print $fh_out_result "week $week_num\n";
    $week_num++;
    my @data;
    open my $fh, '<', $test_file;
    while (<$fh>) {
        chomp;
        my @F = split /,/;
        push @data, [@F];
    }
    close $fh;
    my $prev_time = 0;
    for my $i (0..$#data) {
        my @F = split/:/, $data[$i]->[6];
        my @coeffs = @F[1..3];
        my $fit = $F[4];
        my $result = (split/:/, $data[$i]->[5])[1];
        if (abs($coeffs[0]) < 3 and $fit > 0.94 and $data[$i]->[0] > $prev_time + $past_width + $future_width) {
            my $j = $i;
            $j-- while $data[$j]->[0] >= $data[$i]->[0] - $past_width and $j > 0;
            $j++;
            next if $i == $j or $past_width / ($i - $j) > 250;
            my $out_file = sprintf("temp/%03d.csv", $count);
            open my $fh_out, '>', $out_file;
            for (; $data[$j]->[0] <= $data[$i]->[0] + $future_width; ++$j) {
                print $fh_out join(",", @{$data[$j]}[0,1]) . "\n";
            }
            close $fh_out;
            print $fh_out_result join(",", @coeffs[1, 2], $result)."\n";
            $prev_time = $data[$i]->[0];
            $count++;
        }
    }
}
close $fh_out_result;

sub usage {
    die "Usage: $0 [options] file\n";
}

File: make_past_data.py
#!/usr/bin/env python3
from configparser import ConfigParser
from collections import defaultdict
import sys
import os
import argparse
import csv
import re

def main(r_squared_values, start_week, end_week, debug):
    cfg = ConfigParser()
    cfg.read("config.ini")
    currency = cfg.get("settings", "currency_pair")

    if start_week is None or end_week is None:
        print(f"Usage: {sys.argv[0]} <start week> <end week>")
        sys.exit(-1)

    test_files = []
    for file in os.listdir(f"{currency}/weekly_past_data"):
        if file.startswith("week_") and file.endswith(".csv"):
            week_num = int(re.search(r"week_(\d{3})", file).group(1))
            if start_week <= week_num <= end_week:
                test_files.append((week_num, os.path.join(f"{currency}/weekly_past_data", file)))

    # Sort the list based on week numbers
    test_files.sort()

    week_num = 1
    debug_file_counter = 1
    os.makedirs(f"{currency}/weekly_digest", exist_ok=True)

    for file in os.listdir(f"{currency}/weekly_digest"):
        if re.match(r"week_\d{3}_\d{8}\.csv", file):
            os.remove(os.path.join(f"{currency}/weekly_digest", file))
    
    for _, test_file in test_files:
        output_filename = os.path.basename(test_file)
        with open(f"{currency}/weekly_digest/{output_filename}", "w") as fh_out_result:
            print(test_file)
            week_num += 1

            data = []
            with open(test_file, "r") as fh:
                for line in csv.reader(fh):
                    line[0] = int(line[0])
                    data.append(line)

            prev_time_dict = defaultdict(lambda: defaultdict(int))

            for i in range(len(data)):
                future_data = data[i][5].split("/")
                coeffs_data = data[i][6].split("/")

                for record, coeffs_record in zip(future_data, coeffs_data):
                    future_width, buy_profit, end_time_buy, sell_profit, end_time_sell = map(float, record.split(":"))
                    future_width = int(future_width)
                    l = list(map(float, coeffs_record.split(":")))
                    past_width = int(l[0])
                    coeffs = l[1:4]
                    fit = l[4]

                    density_checked = False
                    density_ok = False
                    for r_squared_value in r_squared_values:
                        if abs(coeffs[0]) < 3 and fit > r_squared_value and data[i][0] > prev_time_dict[past_width][r_squared_value] + past_width + future_width:
                            if not density_checked:
                                density_checked = True
                                j = i
                                while data[j][0] >= data[i][0] - past_width and j > 0:
                                    j -= 1
                                j += 1

                                if i == j or past_width / (i - j) > 250:
                                    density_ok = False
                                    continue
                                else:
                                    density_ok = True
                            else:
                                if not density_ok:
                                    continue
                                

                            if debug:
                                temp_file = f"temp/{debug_file_counter:03d}.csv"
                                with open(temp_file, "w") as temp_csv:
                                    temp_csv_writer = csv.writer(temp_csv)
                                    for row in data[j:i+future_width+1]:
                                        temp_csv_writer.writerow([row[0], row[1]])
                                debug_file_counter += 1

                            fh_out_result.write(f"{data[i][0]},{past_width},{r_squared_value},{coeffs[1]},{coeffs[2]},{int(buy_profit)},{int(sell_profit)}\n")
                            prev_time_dict[past_width][r_squared_value] = data[i][0]

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--r_squared_values", type=float, nargs='+', default=[0.94], help="list of r squared values")
    parser.add_argument("start_week", type=int, nargs='?', default=None, help="start week")
    parser.add_argument("end_week", type=int, nargs='?', default=None, help="end week")
    parser.add_argument("--debug", action="store_true", help="write debug files to temp directory")
    args = parser.parse_args()
    main(args.r_squared_values, args.start_week, args.end_week, args.debug)
